Fixes locking issues when creating channel / adding funds

dirkmc opened the pull request:
Currently when creating a channel or adding funds to an existing channel, the payment channel manager locks up until the messages have been confirmed on chain, which can take several minutes.

This PR resolves the locking issue:
- Operations on each channel are executed sequentially
- Operations on different channels can run in parallel

This PR also manages the state of requests and restores them if lotus is shut down and restarted:
- On startup checks for pending create channel / add funds messages that have been sent to chain
- Writes state to the datastore each time there's a state change
- Keeps track of channels by from / to instead of by channel address, so that state can be written even before channel has been created

This PR also
- sets a flag on channels that are in the "settling" state
- creates a new channel if there is a subsequent call to add funds to that From and To

dirkmc commented on api/apistruct/struct.go:
```diff
@@ -183,7 +183,8 @@ type FullNodeStruct struct {
 
-		PaychGet                   func(ctx context.Context, from, to address.Address, ensureFunds types.BigInt) (*api.ChannelInfo, error)   `perm:"sign"`
+		PaychGet                   func(ctx context.Context, from, to address.Address, amt types.BigInt) (*api.ChannelInfo, error)           `perm:"sign"`
+		PaychGetWaitReady          func(context.Context, cid.Cid) (address.Address, error)                                                   `perm:"sign"` // TODO: is perm:"sign" correct?
```
Should PaychGetWaitReady() have permission `perm:"sign"` ?

magik6k replied:
Yeah, I'd keep it as `sign` for now. We need to rework those anyways

dirkmc commented on node/builder.go:
```diff
@@ -274,6 +275,7 @@ func Online() Option {
 
 			Override(new(*paychmgr.Store), paychmgr.NewStore),
 			Override(new(*paychmgr.Manager), paychmgr.NewManager),
+			Override(HandlePaymentChannelManagerKey, paychmgr.HandleManager),
```
On startup we need to restart tracking of payment channel add funds requests that have already been sent to chain

dirkmc commented on node/impl/paych/paych.go:
```diff
@@ -108,24 +112,7 @@ func (a *PaychAPI) PaychStatus(ctx context.Context, pch address.Address) (*api.P
 }
 
 func (a *PaychAPI) PaychSettle(ctx context.Context, addr address.Address) (cid.Cid, error) {
-
-	ci, err := a.PaychMgr.GetChannelInfo(addr)
-	if err != nil {
-		return cid.Undef, err
-	}
-
-	msg := &types.Message{
-		To:     addr,
-		From:   ci.Control,
-		Value:  types.NewInt(0),
-		Method: builtin.MethodsPaych.Settle,
-	}
-	smgs, err := a.MpoolPushMessage(ctx, msg)
-
-	if err != nil {
-		return cid.Undef, err
-	}
-	return smgs.Cid(), nil
```
This code has moved inside the manager

dirkmc replied:
https://github.com/filecoin-project/lotus/pull/2955

dirkmc commented on paychmgr/manager.go:
```diff
@@ -0,0 +1,278 @@
+// and then lotus was shut down or crashed before the result was
+// received.
+func (pm *Manager) Start(ctx context.Context) error {
+	pm.ctx, pm.shutdown = context.WithCancel(ctx)
```
Is this correct - should we shutdown when the context passed to Start() is cancelled?

magik6k replied:
If this is passed to `fx.Lifecycle`, this context is only alive for the startup process (~5s), you probably want to construct a context in the constructor from `mctx helpers.MetricsCtx, lc fx.Lifecycle`, using `ctx := helpers.LifecycleCtx(mctx, lc)`

dirkmc commented on paychmgr/paych.go:
```diff
@@ -311,35 +270,32 @@ func (pm *Manager) AddVoucher(ctx context.Context, ch address.Address, sv *paych
+	return ca.store.VouchersForPaych(ch)
 }
 
-func (pm *Manager) OutboundChanTo(from, to address.Address) (address.Address, error) {
```
OutboundChanTo() wasn't being used anywhere so I removed it

dirkmc commented on paychmgr/paych.go:
```diff
@@ -355,3 +311,109 @@ func (pm *Manager) NextNonceForLane(ctx context.Context, ch address.Address, lan
+
+// laneState gets the LaneStates from chain, then applies all vouchers in
+// the data store over the chain state
+func (ca *channelAccessor) laneState(state *paych.State, ch address.Address) (map[uint64]*paych.LaneState, error) {
```
laneState() and totalRedeemedWithVoucher() were just moved from state.go

dirkmc commented on paychmgr/state.go:
```diff
@@ -72,80 +73,3 @@ func nextLaneFromState(st *paych.State) uint64 {
-
-// laneState gets the LaneStates from chain, then applies all vouchers in
-// the data store over the chain state
-func (pm *Manager) laneState(state *paych.State, ch address.Address) (map[uint64]*paych.LaneState, error) {
```
laneState() and totalRedeemedWithVoucher() were just moved to paych.go

magik6k commented on paychmgr/manager.go:
```diff
@@ -0,0 +1,278 @@
+		sm:       sm,
+		sa:       &stateAccessor{sm: sm},
+		channels: make(map[string]*channelAccessor),
+		// TODO: Is this the correct way to do this or can I do something different
+		// with dependency injection?
+		pchapi: &paychAPIImpl{api.MpoolAPI, api.StateAPI},
```
I think you should be able to just pass `api` here

magik6k commented on paychmgr/simple.go:
```diff
@@ -17,7 +22,211 @@ import (
+		return
+	}
+
+	head := ca.fundsReqQueue[0]
+	res := ca.processTask(head.ctx, head.from, head.to, head.amt, head.onComplete)
```
I'm pretty sure this was already discussed - but it feels like it should be fairly easy to 'merge' multiple tasks together, so if we need to add funds to 10 different lanes, we'd still only have to block once.

(If this is more than ~15min of work, probably do as a followup PR)

dirkmc replied:
I agree that should be easy to do - I suggest doing it in a follow up PR so we can get this one in today

dirkmc commented:
@magik6k I was able to repro the issues you were seeing when you ran 100 tests. There was a sync issue in the mocks for mempool / state wait that I believe was causing the issues, I've fixed it now and it works for me.
Separately I noticed a couple of missing locks so I added those as well.

Note that it's ok for the tests to output error logs - some of the tests are designed to trigger error cases and then test that the paych manager handles errors correctly.

Please take another look and let me know if it works for you now.

whyrusleeping commented on paychmgr/settle_test.go:
```diff
@@ -0,0 +1,72 @@
+	_, mcid, err := mgr.GetPaych(ctx, from, to, amt)
+	require.NoError(t, err)
+
+	time.Sleep(10 * time.Millisecond)
```
these sleeps make me nervous. If we start seeing random test failures, we're gonna have to do something different

hannahhoward replied:
+1 on this -- I'd like to see this fixed either way in a future PR -- I prefer no time.Sleep's outside of integration tests. It looks like what you'd want to do is write to channels in the mockPaychAPI methods so you can simply delay until they are called.

dirkmc replied:
Resolved by https://github.com/filecoin-project/lotus/pull/2869

whyrusleeping commented on paychmgr/simple.go:
```diff
@@ -41,106 +250,268 @@ func (pm *Manager) createPaych(ctx context.Context, from, to address.Address, am
 	if err != nil {
-		pm.store.lk.Unlock()
-		return address.Undef, cid.Undef, xerrors.Errorf("findChan: %w", err)
+		log.Errorf("Error reading channel info from store: %s", err)
```
will OutboundActiveByFromTo always return a channel info, even if it errors? If not, we need to return here, or something

dirkmc replied:
So I think actually the correct thing to do here is to panic - because if the datastore is not working the system is in an undefined state

dirkmc replied:
https://github.com/filecoin-project/lotus/pull/2870

dirkmc replied:
Magik pointed out that panicking is worse than just bailing: https://github.com/filecoin-project/lotus/pull/2871

hannahhoward commented on paychmgr/simple.go:
```diff
@@ -41,106 +250,268 @@ func (pm *Manager) createPaych(ctx context.Context, from, to address.Address, am
+}
+
+// Change the state of the channel in the store
+func (ca *channelAccessor) mutateChannelInfo(from address.Address, to address.Address, mutate func(*ChannelInfo)) {
```
I think it may be worth looking at [go-statestore](github.com/filecoin-project/go-statestore) even if you're not using `go-statemachine` cause it provides some of this functionality in a simple form. (for the future)

hannahhoward commented on paychmgr/msglistener.go:
```diff
@@ -0,0 +1,61 @@
+	"github.com/ipfs/go-cid"
+)
+
+type msgListener struct {
```
non-blocking suggestion -- you may want to check out [go-pubsub](github.com/hannahhoward/go-pubsub) which is basically this code generalized -- we use it in `go-fil-markets`, `go-datatransfer`, and `go-graphsync`

We have rewritten really simple event emitters like this so many times it'd be nice if possible to not have one more.

dirkmc replied:
haha yeah I don't know why I wrote another pubsub :)

dirkmc replied:
https://github.com/filecoin-project/lotus/pull/2962

hannahhoward commented on paychmgr/paych.go:
```diff
@@ -238,22 +197,22 @@ func (pm *Manager) CheckVoucherSpendable(ctx context.Context, ch address.Address
+	ci, err := ca.store.ByAddress(ch)
 	if err != nil {
-		return types.NewInt(0), err
+		return types.BigInt{}, err
```
Does Lotus still have a BigInt type of its own? I feel like at this point all our money types should come from `spec-actors` abi package? Future refactor.

magik6k replied:
It's an alias of the spec type, we just never got around to getting rid of it

hannahhoward commented on node/impl/paych/paych.go:
```diff
@@ -40,6 +40,10 @@ func (a *PaychAPI) PaychGet(ctx context.Context, from, to address.Address, ensur
 	}, nil
 }
 
+func (a *PaychAPI) PaychGetWaitReady(ctx context.Context, mcid cid.Cid) (address.Address, error) {
```
Possible suggestion: this should take a ChannelID as well, or the whole ChannelInfo struct.

The reason I'd suggest for that is that if I got a message CID from someone else -- but not one I created through PaychGet, I might call this thinking it will work, but absent that message being in the database it won't. If I require a channelID / ChannelInfo I know I need then to reference a message cid I  got from paych Get

hannahhoward replied:
alternatively, just use a UUID instead?

hannahhoward replied:
by this I mean maybe api.Channelnfo has a ChannelMessage that is a UUID, not an actual message CID, and oyu put the message CID int the datastore -- that would prevent people using external message CIDs with this call and also prevent using StateWaitMessage with the channelInfo.ChannelMessage

dirkmc replied:
These are good points 👍 

I think the issue is really a UX issue: if we return a message CID, the client might figure they can use a message CID from someone else when calling PaychGetWaitReady().

My suggestion would be to return a value called "waitID" that is just a stringified version of the message CID. That way the client doesn't think it's a message CID, but we can still store a single value (instead of having to store a separate wait ID).

hannahhoward replied:
sounds good.

dirkmc replied:
https://github.com/filecoin-project/lotus/pull/2986

hannahhoward commented on paychmgr/simple.go:
```diff
@@ -17,7 +22,211 @@ import (
+// address and the CID of the new add funds message.
+// If an operation returns an error, subsequent waiting operations will still
+// be attempted.
+func (ca *channelAccessor) getPaych(ctx context.Context, from, to address.Address, amt types.BigInt) (address.Address, cid.Cid, error) {
```
can I check my understanding on a couple of sequences, based on my read of this code:

I call getPaych between Alice & Bob with $100 -> will return no address, createCID, nil
I immediate call getPaych between Alice & Bob with $100 again -> will block untill creation is done, return address, addCID, nil
I immediately call addVoucher w/ the address returned from call two -- will still work immediately. I only have to wait for AddCid if I care about the second $100 being in the channel.

dirkmc replied:
That's correct 👍

dirkmc commented on paychmgr/manager.go:
```diff
@@ -103,42 +101,9 @@ func HandleManager(lc fx.Lifecycle, pm *Manager) {
-// received.
+// Start restarts tracking of any messages that were sent to chain.
 func (pm *Manager) Start() error {
-	cis, err := pm.store.WithPendingAddFunds()
```
Note: I just moved this into simple.go as it makes more sense to have it there

dirkmc commented on paychmgr/manager.go:
```diff
@@ -0,0 +1,279 @@
+	shutdown context.CancelFunc
+
+	store  *Store
+	sm     StateManagerApi
```
https://github.com/filecoin-project/lotus/pull/2985
